### PR TITLE
Optimize `regexp_replace` when the input is a sparse array

### LINKED ISSUE #3803: Leverage input array's null buffer for regex replace to optimize sparse arrays
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
We can repurpose the input string array's null buffer directly on the specialized regex implementation so that we won't be spending time on re-building it.

**Describe the solution you'd like**
When we know only one input is an array (on the specialized case), we should be able to re-use the existing null buffer. The generic case can also be done in a similar fashion but that would be a bit more hard since now we would need to re-combine 4 null buffers (which might be easy with the underlying bitmap representation, though I doubt it would worth the complexity since the specialized path already covers the generic case).

**Describe alternatives you've considered**
None.

**Additional context**
Originally from @Dandandan as part of #3518.

### PR DESCRIPTION
**Pull request by isidentical**
# Which issue does this PR close?

Closes #3803.

 # Rationale for this change

This is something that arrow-rs already does very heavily, although they also combine it with the unsafe array building APIs which this PR deliberately avoids (reasoning is at the end).

It doesn't provide as much speed-up as the other optimizations for the `regex_replace`, but it seems like a relatively straightforward change for a not-so-bad gain (see benchmarks section). It is also the final optimization candidate (at least localized to the `regex_replace` itself), if I am not missing anything. 

# What changes are included in this PR?

Instead of rebuilding the string array from the iterator, we now build the underlying array data by ourselves in which case we also have the ability to leverage existing null buffers for the input. In the generic version, this couldn't be done without recombining the underlying bitmaps of all the inputs but since this is the specialized case we know for a fact that the only array input is the `strings` (the first argument).

# Are there any user-facing changes?

This is an optimization.

# Benchmarks

As expected, there is no speed-up (or slowdowns, which I guess is noteworthy) in cases where the input array is very dense with data (low amount of NULLs). But depending on the input's data density, we see an average speed-up of %20.

|                            | master | this pr | this pr (unsafe) | speed-up                                             |
|----------------------------|--------|---------|------------------|------------------------------------------------------|
| %100 null & %0 data        | 0.121  | 0.088   | 0.071            | 1.38x against this PR (1.71x against unsafe version) |
| %80 null & %20 data        | 0.489  | 0.396   | 0.347            | 1.23x against this PR (1.40x against unsafe version) |
| %50 null & %50 data        | 0.741  | 0.640   | 0.619            | 1.15x against this PR (1.20x against unsafe version) |
| data > null (80-20, 100-0) | 1.104  | 1.101   | 1.067            | No change (<=%3, within noise)                       |

(Built with the release mode, for a variation of the query/dataset in my [example PR](https://github.com/isidentical/arrow-datafusion/pull/3)).

I have also included a column which indicates the speed-ups when we use the unsafe version but I am not so sure if it makes sense to increase the number of 'unsafe' spots in datafusion for a not-so-crazy speed-up (even if that from the soundness perspective it is safe). I'd be happy to also change my PR to include the unsafe version ([which looks like this](https://github.com/isidentical/arrow-datafusion/pull/3)) if it is not that bad.

### CODE REVIEW

File: datafusion/physical-expr/src/regex_expressions.rs [hunk 1]
```diff
@@ -254,13 +255,38 @@ fn _regexp_replace_static_pattern_replace<T: OffsetSizeTrait>(
...
+        new_offsets.append(T::from_usize(vals.len()).unwrap());
+    });
+
+    let data = ArrayData::try_new(
+        GenericStringArray::<T>::DATA_TYPE,
+        string_array.len(),
+        string_array
+            .data_ref()
+            .null_buffer()
+            .map(|b| b.bit_slice(string_array.offset(), string_array.len())),
```
Comment (alamb) [inline datafusion/physical-expr/src/regex_expressions.rs:283]: 👍  for handling the offset

File: datafusion/physical-expr/src/regex_expressions.rs [hunk 2]
```diff
@@ -513,4 +539,40 @@ mod tests {
...
+
+        let re = _regexp_replace_static_pattern_replace::<i32>(&[
+            Arc::new(values),
+            Arc::new(patterns),
+            Arc::new(replacements),
+        ])
+        .unwrap();
+
+        assert_eq!(re.as_ref(), &expected);
+        assert_eq!(re.null_count(), 4);
```
Comment (alamb) [inline datafusion/physical-expr/src/regex_expressions.rs:576]: To test the offset slicing, I also recommend you run the same test with sliced inputs. Something like

```rust
let values = values.slice(2, 5);
let expected = StringArray::from(vec![
            Some("b"),
            None,
            Some("foo"),
            None,
        ]);


        assert_eq!(re.null_count(), 2);
        let re = _regexp_replace_static_pattern_replace::<i32>(&[
            Arc::new(values),
            Arc::new(patterns),
            Arc::new(replacements),
        ])
        .unwrap();
```

**Review by alamb — APPROVED** [review]
Code looks good to me -- I suggest one more test but otherwise 👍

File: datafusion/physical-expr/src/regex_expressions.rs (hunk 2 above)
Comment (isidentical) [reply datafusion/physical-expr/src/regex_expressions.rs:576]: Great catch, and thanks for the suggestion @alamb! Should be added now.

**Review by alamb — APPROVED** [review]
Thanks @isidentical

**Pull request merged**